src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs

```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+                ? await userManager.GenerateChangeEmailTokenAsync(user, email)
+                : await userManager.GenerateEmailConfirmationTokenAsync(user);
+            code = WebEncoders.Base64UrlEncode(Encoding.UTF8.GetBytes(code));
+
+            var userId = await userManager.GetUserIdAsync(user);
+            var routeValues = new RouteValueDictionary()
+            {
+                ["userId"] = userId,
+                ["code"] = code,
+            };
+
+            if (isChange)
+            {
+                // This is validated by the /confirmEmail endpoint on change.
+                routeValues.Add("changedEmail", email);
+            }
+
+            var confirmEmailUrl = linkGenerator.GetPathByName(confirmEmailEndpointName, routeValues)
+                ?? throw new NotSupportedException($"Could not find endpoint named '{confirmEmailEndpointName}'.");
+
+            await emailSender.SendEmailAsync(email, "Confirm your email",
+                $"Please confirm your account by <a href='{HtmlEncoder.Default.Encode(confirmEmailUrl)}'>clicking here</a>.");
+        }
+
         return new IdentityEndpointsConventionBuilder(routeGroup);
     }
 
+    private static ValidationProblem CreateValidationProblem(string errorCode, string errorDescription) =>
+        TypedResults.ValidationProblem(new Dictionary<string, string[]> {
+            { errorCode, new[] { errorDescription } }
+        });
+
+    private static ValidationProblem CreateValidationProblem(params IdentityResult[] results)
+    {
+        var errorDictionary = new Dictionary<string, string[]>(results.Length);
+
+        foreach (var result in results)
+        {
+            Debug.Assert(!result.Succeeded);
+            foreach (var error in result.Errors)
+            {
+                string[] newDescriptions;
+
+                if (errorDictionary.TryGetValue(error.Code, out var descriptions))
+                {
+                    newDescriptions = new string[descriptions.Length + 1];
+                    Array.Copy(descriptions, newDescriptions, descriptions.Length);
+                    newDescriptions[descriptions.Length] = error.Description;
+                }
+                else
+                {
+                    newDescriptions = new[] { error.Description };
+                }
+
+                errorDictionary[error.Code] = newDescriptions;
+            }
+        }
+
+        return TypedResults.ValidationProblem(errorDictionary);
+    }
+
+    private static async Task<TwoFactorResponse> CreateTwoFactorResponseAsync<TUser>(TUser user, SignInManager<TUser> signInManager, string[]? recoveryCodes = null)
+        where TUser : class
+    {
+        var userManager = signInManager.UserManager;
+
+        var key = await userManager.GetAuthenticatorKeyAsync(user);
+        if (string.IsNullOrEmpty(key))
+        {
+            await userManager.ResetAuthenticatorKeyAsync(user);
+            key = await userManager.GetAuthenticatorKeyAsync(user);
```

Comment:
Can this still be `null` the second time?

Reply:
The API signature says yes but ResetAuthenticatorKeyAsync succeeding says no. We should probably validate and actually check the IdentityResult returned by ResetAuthenticatorKeyAsync too. This does match the Identity UI behavior though.